{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "cell_id": "00000-54378b4b-89fc-42b2-9ee3-66b1d1ba47b0",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 21475,
    "execution_start": 1635545408316,
    "source_hash": "fec3ce2a",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "'''\n",
    "Function created to take in GenBank File (as a list) and get data frame.\n",
    "Function iterates over objects (sequences in the genome) in the file \n",
    "to read their features. If the feature type is a CDS (if the sequence \n",
    "is a coding region) then all of the qualifiers of that CDS are accessed\n",
    "and put into a dictionary. The contig ID and contig length are assigned \n",
    "keys in the dictionary. \n",
    "\n",
    "'''\n",
    "\n",
    "def get_df(recs):\n",
    "    assert isinstance(recs, list), \"Need a list as input\"\n",
    "    assert isinstance(recs[0], SeqRecord), \"The input list should contain SeqRecord objects\"\n",
    "    fdicts = []\n",
    "    for r in recs:\n",
    "        for feature in r.features:\n",
    "            if feature.type==\"CDS\":\n",
    "                d = dict(feature.qualifiers)\n",
    "                for k,v in d.items():\n",
    "                    if len(v)==1:\n",
    "                        d[k] = v[0]\n",
    "                d['contig_id'] = r.id\n",
    "                d['contig_len'] = len(str(r.seq))\n",
    "                fdicts.append(d)\n",
    "    dfa = pd.DataFrame(fdicts) \n",
    "    return dfa\n",
    "\n",
    "'''\n",
    "The GenBank files for A. Nidulans and A. tubingensis are made into lists\n",
    "and assigned to variable names.\n",
    "The list of the Genbank File for each organism is passed into the\n",
    "function above to get the dataframe. The organism key in the dictionary\n",
    "created is assigned to the organism's name. \n",
    "'''\n",
    "\n",
    "r_nidulans = list(SeqIO.parse(\"aspergillus_nidulans.gbk\", \"genbank\"))\n",
    "df_nidulans = get_df(r_nidulans)\n",
    "df_nidulans['org'] = \"A. nidulans\"\n",
    "\n",
    "r_tubingensis = list(SeqIO.parse(\"aspergillus_tubingensis.gbk\", \"genbank\"))\n",
    "df_tubingensis = get_df(r_tubingensis)\n",
    "df_tubingensis['org'] = \"A. tubingensis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "cell_id": "00001-2f368300-b9be-464b-83b9-e7b7a0b58783",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 37,
    "execution_start": 1635545429806,
    "source_hash": "42627451",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30200029"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Df_all is assigned to both the data frame variables above for each org.\n",
    "Df_nidulans is assigned to the number of proteins in the A. nidulans\n",
    "dataframe. Df_tmp fgets the number of unique contigs in the genome \n",
    "and then is used to get the genome size.\n",
    "\"\"\"\n",
    "df_all = pd.concat([df_nidulans, df_tubingensis])\n",
    "df_nidulans['protein_id'].size \n",
    "df_tmp = df_nidulans[['contig_id', 'contig_len']].drop_duplicates() \n",
    "df_tmp['contig_len'].sum() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "cell_id": "00002-06725818-8b1d-4e0e-8def-ec384b52350c",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 71,
    "execution_start": 1635545429845,
    "source_hash": "f666ff31",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#df_all gets the proteins created by each contig.\n",
    "df_all['protein_count_by_contig'] = df_all.groupby('contig_id')['protein_id'].transform(pd.Series.nunique)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "cell_id": "00003-33a3a4b9-67c5-432e-bc2a-b1a780a25962",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 0,
    "execution_start": 1635545429968,
    "source_hash": "d446f3f1",
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "df_subset creates another dataframe for both organisms together, and\n",
    "contains info about which org it is, what contigs are present, \n",
    "the length of each unique contig, and how many proteins are made from \n",
    "each contig.\n",
    "\"\"\"\n",
    "df_subset = df_all[\n",
    "    ['org', 'contig_id',\n",
    "    'contig_len',\n",
    "    'protein_count_by_contig']\n",
    "    ].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "cell_id": "00004-45827667-82c7-424d-a6a2-f32b165ba6c6",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 11,
    "execution_start": 1635545429995,
    "source_hash": "525687f0",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>contig_len</th>\n",
       "      <th>protein_count_by_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NT_107008.1</td>\n",
       "      <td>1482181</td>\n",
       "      <td>486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NT_107009.1</td>\n",
       "      <td>2200800</td>\n",
       "      <td>689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1175</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NW_101435.1</td>\n",
       "      <td>8226</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NT_107005.1</td>\n",
       "      <td>1394314</td>\n",
       "      <td>445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1625</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NT_107012.1</td>\n",
       "      <td>2569337</td>\n",
       "      <td>818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9291</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NW_023336285.1</td>\n",
       "      <td>2136001</td>\n",
       "      <td>755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10046</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NW_023336286.1</td>\n",
       "      <td>798263</td>\n",
       "      <td>259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10305</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NW_023336287.1</td>\n",
       "      <td>1448135</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10765</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NW_023336288.1</td>\n",
       "      <td>231917</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10830</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NW_023336289.1</td>\n",
       "      <td>1979743</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  org       contig_id  contig_len  protein_count_by_contig\n",
       "0         A. nidulans     NT_107008.1     1482181                      486\n",
       "486       A. nidulans     NT_107009.1     2200800                      689\n",
       "1175      A. nidulans     NW_101435.1        8226                        5\n",
       "1180      A. nidulans     NT_107005.1     1394314                      445\n",
       "1625      A. nidulans     NT_107012.1     2569337                      818\n",
       "...               ...             ...         ...                      ...\n",
       "9291   A. tubingensis  NW_023336285.1     2136001                      755\n",
       "10046  A. tubingensis  NW_023336286.1      798263                      259\n",
       "10305  A. tubingensis  NW_023336287.1     1448135                      460\n",
       "10765  A. tubingensis  NW_023336288.1      231917                       65\n",
       "10830  A. tubingensis  NW_023336289.1     1979743                      647\n",
       "\n",
       "[92 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Calling upon the variable df_subset gives a chart representing\n",
    "the information discussed in the cell above.\n",
    "\"\"\"\n",
    "df_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "cell_id": "00005-cbc6826d-1096-4100-88a4-c2c1888ce781",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 6,
    "execution_start": 1635545430043,
    "source_hash": "8b6dcebc",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_proteins</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>total_contig</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>org</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A. nidulans</th>\n",
       "      <td>9556</td>\n",
       "      <td>30200029</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A. tubingensis</th>\n",
       "      <td>11477</td>\n",
       "      <td>35047229</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                all_proteins  genome_size  total_contig\n",
       "org                                                    \n",
       "A. nidulans             9556     30200029            77\n",
       "A. tubingensis         11477     35047229            15"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "This code gets the number of proteins total, the total length of \n",
    "all the contigs, and how many contigs are present for both organisms.\n",
    "It then puts this information into a table. \n",
    "\"\"\"\n",
    "df_subset.groupby('org').agg(\n",
    "    all_proteins=('protein_count_by_contig', sum),\n",
    "    genome_size=('contig_len' , sum),\n",
    "    total_contig=('contig_id' , len)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "cell_id": "00006-a0d7576f-ca99-4ee3-83c6-b37186ede27d",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 12611,
    "execution_start": 1635545430058,
    "source_hash": "33d0efd8",
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This function gets the biosyntethic gene cluster information.\n",
    "A list (GenBank File converted to list) is read and if an object \n",
    "of a feature is a protocluster, then the product of the protocluster\n",
    "is assigned to the variable tmp. The tmp is then added into a list keeping\n",
    "track of the protocluster product's or an organism.\n",
    "\"\"\"\n",
    "def get_bgc_df(records, o):\n",
    "    results = []\n",
    "    for r in records:\n",
    "        for f in r.features:\n",
    "            if f.type == 'protocluster':\n",
    "                if f.qualifiers:\n",
    "                    tmp = f.qualifiers.get('product')\n",
    "                    if tmp:\n",
    "                        tmp = tmp[0]\n",
    "                        results.append({'org': o, 'product': tmp})\n",
    "\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "'''\n",
    "The Genbank files of A. tubingensis are made into a list and assigned\n",
    "to a variable. This variable is then passed through the function \n",
    "above to get the dataframe describing the protoclusters. The same is \n",
    "done for the Genbank file on A. nidulans.\n",
    "'''\n",
    "tubingensis = list(SeqIO.parse(\n",
    "    \"aspergillus_tubingensis.gbk\", \"genbank\")\n",
    "    )\n",
    "df_tubingensis = get_bgc_df(tubingensis, 'A. tubingensis')\n",
    "\n",
    "nidulans = list(SeqIO.parse(\n",
    "    \"aspergillus_nidulans.gbk\", \"genbank\")\n",
    "    )\n",
    "df_nidulans = get_bgc_df(nidulans, 'A. nidulans')\n",
    "\n",
    "bgc_df = pd.concat([df_tubingensis, df_nidulans])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "cell_id": "00007-a0305db6-3de0-4b54-8da8-9ad431270b8f",
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 104,
    "execution_start": 1635545533402,
    "source_hash": "dbd60b89",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org</th>\n",
       "      <th>product</th>\n",
       "      <th>product_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NRPS</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>NRPS-like</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>T1PKS</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>betalactone</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>indole</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A. nidulans</td>\n",
       "      <td>terpene</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NRPS</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>NRPS-like</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>T1PKS</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>T3PKS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>betalactone</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>indole</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>siderophore</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A. tubingensis</td>\n",
       "      <td>terpene</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               org      product  product_count\n",
       "0      A. nidulans         NRPS             13\n",
       "1      A. nidulans    NRPS-like             11\n",
       "2      A. nidulans        T1PKS             23\n",
       "3      A. nidulans  betalactone              3\n",
       "4      A. nidulans       indole              5\n",
       "5      A. nidulans      terpene              9\n",
       "6   A. tubingensis         NRPS             20\n",
       "7   A. tubingensis    NRPS-like             20\n",
       "8   A. tubingensis        T1PKS             39\n",
       "9   A. tubingensis        T3PKS              1\n",
       "10  A. tubingensis  betalactone              1\n",
       "11  A. tubingensis       indole              2\n",
       "12  A. tubingensis  siderophore              2\n",
       "13  A. tubingensis      terpene             21"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "bgc_overview is a variable that contains information about each organism and the protocluster products. \n",
    "It contains the protein type, and how many of each are made. \n",
    "\"\"\"\n",
    "bgc_overview = bgc_df.groupby(['org', 'product']).agg(product_count=('product', len)).reset_index()\n",
    "\n",
    "bgc_overview"
   ]
  }
 ],
 "metadata": {
  "deepnote": {
   "is_reactive": false
  },
  "deepnote_execution_queue": [],
  "deepnote_notebook_id": "7118dacd-12d5-413f-8907-4b856ee99835",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
